fuzzing/bls12381.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2020 Martin Holst Swende, Marius van der Wijden
+// This file is part of the goevmlab library.
+//
+// The library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the goevmlab library. If not, see <http://www.gnu.org/licenses/>.
+
+package fuzzing
+
+import (
+	"fmt"
+	"io/ioutil"
+	"math/big"
+	"math/rand"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/holiman/goevmlab/ops"
+	"github.com/holiman/goevmlab/program"
+)
+
+var precompilesBLS = []int{
+	10, 11, 12, 13, 14, 15, 16, 17, 18,
+}
+
+var inputSizes = []int{
+	256, 256, 160, 512, 512, 288, 384, 64, 128,
+}
+
+var ouputSizes = []int{
+	128, 128, 128, 256, 256, 256, 32, 128, 256,
+}
+
+func GenerateBLS() *GstMaker {
+	gst := basicStateTest("Yolo-Ephnet-1")
+	// Add a contract which calls BLS
+	dest := common.HexToAddress("0x00ca110b15012381")
+	gst.AddAccount(dest, GenesisAccount{
+		Code:    RandCallBLS(),
+		Balance: new(big.Int),
+		Storage: make(map[common.Hash]common.Hash),
+	})
+	// The transaction
+	{
+		tx := &stTransaction{
+			// 8M gaslimit
+			GasLimit:   []uint64{8000000},
+			Nonce:      0,
+			Value:      []string{randHex(4)},
+			Data:       []string{randHex(100)},
+			GasPrice:   big.NewInt(0x01),
+			To:         dest.Hex(),
+			PrivateKey: hexutil.MustDecode("0x45a915e4d060149eb4365960e6a7a45f334393093061116b197e3240065ff2d8"),
+		}
+		gst.SetTx(tx)
+	}
+	return gst
+}
+
+func RandCallBLS() []byte {
+	p := program.NewProgram()
```

Comment:
One thing you _could_ do, is make this shim contract call each one of the precomples, serially, and store their output in slots [0 .. 9].

Reply:
I tried, I'm not sure about my offset calculation, PTAL